src/api/app/models/worker_status.rb

```diff
@@ -44,88 +44,83 @@ def project_names(prjs)
-    @squeues = {}
-    Rails.cache.write('workerstatus', ret, expires_in: 3.minutes)
+    @squeues = Hash.new(0)
+
+    Rails.cache.write('workerstatus', backend_status, expires_in: 3.minutes)
+
     StatusHistory.transaction do
-      wdata.elements('blocked') do |e|
-        save_value_line(e, 'blocked')
-      end
-      wdata.elements('waiting') do |e|
-        save_value_line(e, 'waiting')
-      end
-      wdata.elements('partition') do |p|
-        p.elements('daemon') do |daemon|
-          parse_daemon_infos(daemon)
-        end
+      ['blocked', 'waiting'].each do |state|
+        wdata.search("//#{state}").each { |e| save_value_line(e, state) }
       end
+      wdata.search('partition/daemon').each { |daemon| parse_daemon_infos(daemon) }
       parse_worker_infos(wdata)
-      @squeues.each_pair do |key, value|
-        StatusHistory.create(time: @mytime, key: key, value: value)
-      end
+      @squeues.each_pair { |key, value| StatusHistory.create(time: @mytime, key: key, value: value) }
     end
-    ret
+
+    backend_status
   end
 
   private
 
   def add_squeue(key, value)
-    @squeues[key] ||= 0
     @squeues[key] += value.to_i
   end
 
   def parse_daemon_infos(daemon)
-    return unless daemon['type'] == 'scheduler'
-    arch = daemon['arch']
-    # FIXME2.5: The current architecture model is a gross hack, not connected at all
-    #           to the backend config.
-    a = Architecture.find_by_name(arch)
-    if a
-      a.available = true
-      a.save
-    end
-    queue = daemon.get('queue')
+    return unless daemon.attributes['type'].value == 'scheduler'
+
+    queue = daemon.at_xpath('queue')
     return unless queue
-    ['high', 'next', 'med', 'low'].each { |key| add_squeue("squeue_#{key}_#{arch}", queue[key]) }
+
+    ['high', 'next', 'med', 'low'].each do |key|
+      s_key = squeue_key([key, daemon_architecture(daemon.attributes['arch'].value).name])
+      add_squeue(s_key, queue.attributes[key].value)
+    end
+  end
+
+  def squeue_key(key_parts)
+    generic_key_generation(key_parts.prepend('squeue'))
+  end
+
+  def generic_key_generation(key_parts)
+    key_parts.join('_')
+  end
+
+  def daemon_architecture(arch)
+    Architecture.find_by_name(arch).tap { |a| a.update(available: true) }
   end
 
   def parse_worker_infos(wdata)
-    allworkers = {}
+    allworkers = Hash.new(0)
     workers = {}
+
     ['building', 'idle', 'dead', 'down', 'away'].each do |state|
-      wdata.elements(state) do |e|
-        id = e['workerid']
-        if workers.key?(id)
-          Rails.logger.debug 'building+idle worker'
-          next
-        end
-        workers[id] = 1
-        key = state + '_' + e['hostarch']
-        allworkers["building_#{e['hostarch']}"] ||= 0
-        allworkers["idle_#{e['hostarch']}"] ||= 0
-        allworkers["dead_#{e['hostarch']}"] ||= 0
-        allworkers["down_#{e['hostarch']}"] ||= 0
-        allworkers["away_#{e['hostarch']}"] ||= 0
-        allworkers[key] = allworkers[key] + 1
```

Comment:
do we need to create a `StatusHistory` for those which weren't found? For example if you run the same spec on master, it will create 3 more `StatusHistory` elements for (dead, down, away) even though, no dead, down or always workers were found (but they are being manually initialized with 0 in the master )

Reply:
You don't, but dropping them makes scaling a lttle more complex - same for plotting. But give it a try

Reply:
ok let's see how it goes, thanks @coolo